pkg/promclient/promclient.go

```diff
@@ -0,0 +1,200 @@
+// Package promclient offers helper client function for various API endpoints.
+
+package promclient
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"path"
+	"path/filepath"
+	"sort"
+	"time"
+
+	"github.com/improbable-eng/thanos/pkg/tracing"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/promql"
+
+	"github.com/go-kit/kit/log"
+	"github.com/improbable-eng/thanos/pkg/runutil"
+	"github.com/pkg/errors"
+	promlabels "github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/tsdb/labels"
+	"gopkg.in/yaml.v2"
+)
+
+// IsWALFileAccesible returns no error if WAL dir can be found. This helps to tell
+// if we have access to Prometheus TSDB directory.
+func IsWALDirAccesible(dir string) error {
+	const errMsg = "WAL dir is not accessible. Is this dir a TSDB directory? If yes it is shared with TSDB?"
+
+	f, err := os.Stat(filepath.Join(dir, "wal"))
+	if err != nil {
+		return errors.Wrap(err, errMsg)
+	}
+
+	if !f.IsDir() {
+		return errors.New(errMsg)
+	}
+
+	return nil
+}
+
+// ExternalLabels returns external labels from /api/v1/status/config Prometheus endpoint.
+// Note that configuration can be hot reloadable on Prometheus, so this config might change in runtime.
+func ExternalLabels(ctx context.Context, logger log.Logger, base *url.URL) (labels.Labels, error) {
+	u := *base
+	u.Path = path.Join(u.Path, "/api/v1/status/config")
+
+	req, err := http.NewRequest(http.MethodGet, u.String(), nil)
+	if err != nil {
+		return nil, errors.Wrap(err, "create request")
+	}
+	resp, err := http.DefaultClient.Do(req.WithContext(ctx))
+	if err != nil {
+		return nil, errors.Wrapf(err, "request config against %s", u.String())
+	}
+	defer runutil.CloseWithLogOnErr(logger, resp.Body, "query body")
+
+	b, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, errors.Errorf("failed to read body")
+	}
+
+	if resp.StatusCode != 200 {
+		return nil, errors.Errorf("got non-200 response code: %v, response: %v", resp.StatusCode, string(b))
+	}
+
+	var d struct {
+		Data struct {
+			YAML string `json:"yaml"`
+		} `json:"data"`
+	}
+	if err := json.Unmarshal(b, &d); err != nil {
+		return nil, errors.Wrapf(err, "unmarshal response: %v", string(b))
+	}
+	var cfg struct {
+		Global struct {
+			ExternalLabels map[string]string `yaml:"external_labels"`
+		} `yaml:"global"`
+	}
+	if err := yaml.Unmarshal([]byte(d.Data.YAML), &cfg); err != nil {
+		return nil, errors.Wrapf(err, "parse Prometheus config: %v", d.Data.YAML)
+	}
+	return labels.FromMap(cfg.Global.ExternalLabels), nil
+}
+
+func QueryInstant(ctx context.Context, logger log.Logger, addr, query string, t time.Time) (promql.Vector, error) {
```

Comment:
Would it be worth changing this signature to be consistent with the `ExternalLabels` one? `base *url.URL` instead of `addr string`?